File: JavaCode/src/com/ysx/lintcode/easy/WordPattern.java
package com.ysx.lintcode.easy;

import java.util.HashMap;
import java.util.Map;

/**
 * Lintcode 828. Word Pattern
 * http://www.lintcode.com/zh-cn/problem/word-pattern/
 * Leetcode 290. Word Pattern
 * https://leetcode.com/problems/word-pattern/description/
 */
public class WordPattern {

    public static void main(String[] args) {
        WordPattern test = new WordPattern();
        String pattern = "abba";
        String str = "dog cat cat dog";
        System.out.println(test.wordPattern(pattern, str));

        String pattern2 = "abba";
        String str2 = "dog dog dog dog";
        System.out.println(test.wordPattern(pattern2, str2));
    }

    /**
     * @param pattern: a string, denote pattern string
     * @param str:     a string, denote matching string
     * @return: an boolean, denote whether the pattern string and the matching string match or not
     */
    public boolean wordPattern(String pattern, String str) {
        String[] strArray = str.split(" ");
        if (pattern.length() != strArray.length) {
            return false;
        }
        Map<Character, String> map = new HashMap<>();
        int length = pattern.length();
        for (int i = 0; i < length; i++) {
            char ch = pattern.charAt(i);
            if (map.containsKey(ch)) {
                if (!map.get(ch).equals(strArray[i])) {
                    return false;
                }
            } else {
                if (map.containsValue(strArray[i])) {
                    return false;
                }
                map.put(ch, strArray[i]);
            }
        }
        return true;
    }
}

File: KotlinCode/src/lintcode/easy/WordPattern.kt
package lintcode.easy

import java.util.HashMap

class WordPattern{
    fun wordPattern(pattern: String, str: String): Boolean {
        val strArray = str.split(" ")
        if (pattern.length != strArray.size) {
            return false
        }
        val length = pattern.length
        val map = HashMap<Char, String>()
        for (i in 0 until length) {
            val ch = pattern[i]
            if (map.containsKey(ch)) {
                if (map[ch] != strArray[i]) {
                    return false
                }
            } else {
                if (map.containsValue(strArray[i])) {
                    return false
                }
                map[ch] = strArray[i]
            }
        }
        return true
    }
}

fun main(args: Array<String>) {
    val test = WordPattern()
    val pattern1 = "abba"
    val str1 = "dog cat cat dog"
    println(test.wordPattern(pattern1, str1))

    val pattern2 = "abba"
    val str2 = "dog dog dog dog"
    println(test.wordPattern(pattern2, str2))
}